src/scanpy/preprocessing/_simple.py

```diff
@@ -612,6 +612,47 @@ def normalize_per_cell(  # noqa: PLR0917
     return X if copy else None
 
 
+@numba.njit(cache=True, parallel=True)
+def to_dense(shape, indptr, indices, data):
+    """\
+    Numba kernel for np.toarray() function
+    """
+    X = np.empty(shape, dtype=data.dtype)
+
+    for r in numba.prange(shape[0]):
+        X[r] = 0
+        for i in range(indptr[r], indptr[r + 1]):
+            X[r, indices[i]] = data[i]
+    return X
+
+
+def numpy_regress_out(
+    X: np.ndarray,
+    A: np.ndarray,
```

Comment:
please use speaking parameter names, in `lower_snake_case`.

`X` is reserved for `AnnData.x`, i.e. the main matrix of an AnnData object.